File: Day_14_Challange_2.py
from typing import List

"""
Creator: Przemysław Szewczak
Date: 27.12.2021
Source: https://adventofcode.com/2021/day/14
Language: Python 3.9
"""


class Polymers:

    def __init__(self):
        self.polymer = ""
        self.polymer_template = {}
        self.pairs = {}
        self._load_data()
        self._create_first_pairs()

    def __call__(self, *args, **kwargs) -> None:
        for _ in range(40):
            self._create_pairs()
        pass

    def __str__(self):
        """Not used in challange 2"""
        print(self.polymer)
        [print(f"{k} - {v}") for k, v in self.polymer_template.items()]
        return ''

    def _load_data(self) -> None:
        """Private load data from file method used in __init__"""
        with open('test_input.txt', 'r') as file:
            self.polymer = file.readline().strip()
            file.readline()
            for line in file:
                k, v = line.strip().split(' -> ')
                self.polymer_template[k] = v
        return None

    def _create_first_pairs(self) -> None:
        """Init creation first set of paris"""
        for starting_ind in range(int(len(self.polymer) - 1)):
            pair = (self.polymer[starting_ind] + self.polymer[(starting_ind + 1)])
            if pair in self.pairs:
                self.pairs[pair] += 1
            else:
                self.pairs[pair] = 1

    def _create_pairs(self) -> None:
        """Create pairs and add them to pair dictionary, removing previous pairs"""
        pair_keys = self.pairs.keys()
        to_add = {}
        for pair in pair_keys:
            first_new_pair = pair[0] + self.polymer_template[pair]
            second_new_pair = self.polymer_template[pair] + pair[1]
            if first_new_pair in to_add.keys():
                to_add[first_new_pair] += self.pairs[pair]
            else:
                to_add[first_new_pair] = self.pairs[pair]
            if second_new_pair in to_add.keys():
                to_add[second_new_pair] += self.pairs[pair]
            else:
                to_add[second_new_pair] = self.pairs[pair]
            if pair in to_add.keys():
                to_add[pair] -= self.pairs[pair]
            else:
                to_add[pair] = (self.pairs[pair]) * (-1)
        for k, v in to_add.items():
            if k in self.pairs.keys():
                self.pairs[k] += v
            else:
                self.pairs[k] = v
            if self.pairs[k] == 0:
                del (self.pairs[k])
        return None

    def _count_letters(self) -> List[int]:
        """Counting letters private method"""
        letters = {}
        for pair in self.pairs:
            if pair[0] in letters:
                letters[pair[0]] += self.pairs[pair]
            else:
                letters[pair[0]] = self.pairs[pair]
        letters[self.polymer[-1]] += 1
        # print(self.pairs)
        return [max(letters.values()), min(letters.values())]

    def return_answer(self) -> int:
        """Answer return method"""
        numbers = self._count_letters()
        return numbers[0] - numbers[1]


def main() -> None:
    polymer = Polymers()
    polymer()
    print(f"Solution is: {polymer.return_answer()}")
    return None


if __name__ == '__main__':
    main()

File: Day_14_Challange_2.cs
using System;
using System.IO;
using System.Collections.Generic;


/*
Creator: Przemysław Szewczak
Date: 27.12.2021
Source: https://adventofcode.com/2021/day/10
Language: C#
*/

namespace Solution
{
    internal class Challange1
    {
        public static void Main()
        {
            Polymers poly = new Polymers();
        }
        
    }
    public class Polymers
    {
        private string startingPolymer = "";
        private Dictionary<string, string> polymerPatterns = new Dictionary<string, string>();
        private Dictionary<string, long> polymerPairs = new Dictionary<string, long>();
        private Dictionary<char, long> letterCount = new Dictionary<char, long>();
        private char lastLetter = ' ';
         public Polymers()
        {
            Read();
            createFirstPairs();
            for (int i = 0; i < 40; i++)
            {
                createNewpairs();
            }
            countLetters();
            long day14Result = result();
            Console.WriteLine($"Day 14 Challenge 2 result is: {day14Result}");
        }
        private void Read()
        {
            // Load data from file method
            int lineCount = 0;
            foreach (string line in File.ReadLines(
                         @"C:\Przemek\Python\PythonAdventCalendar\AdventOfCode\test_input.txt"))
            {
                if (lineCount == 0)
                {
                    startingPolymer = line.Trim();
                    lineCount += 1;
                    continue;
                }
                else if (lineCount == 1)
                {
                    lineCount += 1;
                    continue;
                }

                string[] pattern = line.Trim().Split(" -> ");
                polymerPatterns.Add(pattern[0], pattern[1]);
            }

        }
        
        public void showPattern()
        {
            // Showing pattern list - Not used in final Program
            Console.WriteLine("Showing all Patterns");
            foreach( KeyValuePair<string, string> dict in polymerPatterns)
            {
                Console.WriteLine($"{dict.Key} -- {dict.Value}");
            }
        }
        
        public void showPairs()
        {
            // Showing actual pairs in polymers - Not used in final Program
            Console.WriteLine("Actual Pairs");
            foreach( KeyValuePair<string,long> kv in polymerPairs)
            {
                Console.WriteLine($"{kv.Key} = {kv.Value}");
            }
        }
        
        private void createFirstPairs()
        {
            // Creating first set of pairs after Read() method was called
            string pair = "";
            char[] polymerChars = startingPolymer.ToCharArray();
            for (int index = 0; index < startingPolymer.Length - 1; index++)
            {
                pair = polymerChars[index].ToString() + polymerChars[index + 1].ToString();
                if (polymerPairs.ContainsKey(pair))
                {
                    polymerPairs[pair] += 1;
                }
                else
                {
                    polymerPairs.Add(pair,1); 
                }
                
            }

            lastLetter = polymerChars[startingPolymer.Length - 1];
        }

        private void createNewpairs()
        {
            // Creates new pairs using actual pair dictionary
            var keys = polymerPairs.Keys;
            Dictionary<string, long> toAddition = new Dictionary<string, long>();
            foreach (string key in keys)
            {
                
                char[] polymerParts = key.ToCharArray();
                string firstPartPolymer = polymerParts[0].ToString() + polymerPatterns[key];
                string secondPartPolymer = polymerPatterns[key] + polymerParts[1].ToString();
                if (toAddition.ContainsKey(firstPartPolymer))
                {
                    toAddition[firstPartPolymer] += polymerPairs[key];
                }
                else
                {
                    toAddition.Add(firstPartPolymer,polymerPairs[key]);
                }
                //
                if (toAddition.ContainsKey(secondPartPolymer))
                {
                    toAddition[secondPartPolymer] += polymerPairs[key];
                }
                else
                {
                    toAddition.Add(secondPartPolymer,polymerPairs[key]);
                }
                //
                if (toAddition.ContainsKey(key))
                {
                    toAddition[key] -= polymerPairs[key];
                }
                else
                {
                    toAddition.Add(key,polymerPairs[key]*(-1));
                }


            }

            foreach (KeyValuePair<string, long> kv in toAddition)
            {
                if (polymerPairs.ContainsKey(kv.Key))
                {
                    polymerPairs[kv.Key] += kv.Value;
                }
                else
                {
                    polymerPairs.Add(kv.Key,kv.Value);
                }

                if (polymerPairs[kv.Key] == 0)
                {
                    polymerPairs.Remove(kv.Key);
                }
            }
        }
        
        private void addNewPair(string pair)
        {
            // Adding new pair to polymer pairs dictionary - Not used in final Program
            // Prototype method used before final solution
            string newPair = "";
            char[] polymerChars = pair.ToCharArray();
            // Console.WriteLine(polymerChars[0]);
            for (int index = 0; index < pair.Length - 1; index++)
            {
                newPair = polymerChars[index].ToString() + polymerChars[index + 1].ToString();
                // polymerPairs.ContainsKey(newPair)? polymerPairs[newPair]+=1 : polymerPairs.Add(pair,1);
                 if (polymerPairs.ContainsKey(newPair))
                 {
                     // Console.WriteLine($"Dict contains {newPair}");
                     polymerPairs[newPair] += 1;
                 }
                 else
                 {
                     // Console.WriteLine($"{newPair} does not exists");
                     polymerPairs.Add(newPair,1);
                 }
                // polymerPairs.Add(pair,1);
            }
        }

        private void countLetters()
        {
            // Counting first letters in pairs and adding last letter.
            foreach (KeyValuePair<string, long> kv in polymerPairs)
            {
                char[] letters = kv.Key.ToCharArray();
                if (letterCount.ContainsKey(letters[0]))
                {
                    letterCount[letters[0]] += kv.Value;
                }
                else
                {
                    letterCount.Add(letters[0],kv.Value);
                }
            }

            letterCount[lastLetter] += 1;
        }

        public void showLetters()
        {
            // Showing letters, used for visualisation - Not used in final program
            foreach (KeyValuePair<char, long> kv in letterCount)
            {
                Console.WriteLine($"{kv.Key}  =  {kv.Value}");
            }
        }

        private long result()
        {
            // Calculating result
            var values = letterCount.Values;
            long maxLetter = values.Max();
            long minLetter = values.Min();
            return maxLetter - minLetter;
        }
    }
}

